sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,319 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "sw/device/lib/base/csr.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/math.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/rand_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#define PERFTEST_MAX_NUM_REPS 20
+#define PERFTEST_BUF_LEN 1000
+
+typedef struct test_context {
+  uint8_t buf1[PERFTEST_BUF_LEN];
+  uint8_t buf2[PERFTEST_BUF_LEN];
+  size_t len;
+} test_context_t;
+
+typedef struct perf_test {
+  // A human-readable name for this particular test, e.g. "memcpy".
+  const char *label;
+
+  // Setup functions for each context buffer. These function pointers may be
+  // NULL. The runtime of these functions will not be measured.
+  void (*setup_buf1)(uint8_t *, size_t);
+  void (*setup_buf2)(uint8_t *, size_t);
+
+  // Points to a function that exercises the function under test. This is the
+  // behavior that should be timed. It must not be NULL.
+  void (*func)(test_context_t *);
+
+  // The number of times to repeat this particular perf test. This is a
+  // per-test option because our 32-bit variance computation is sensitive to
+  // numerical overflow; tests that have a high variance should be run fewer
+  // times to avoid this.
+  size_t num_reps;
```

Comment:
IMHO we can leave this out since at this level we don't expect a lot of variance, do we?

Reply:
Well, it depends on the nature of the perftest, which I haven't quite nailed down yet. Currently memcpy, memset, memcmp, and memrcmp have zero variance. The others have too much to fit in a u32, so I can't easily print the results with `base_printf()`! For instance, the memchr perftest randomizes `buf1` and `buf2` and then searches `buf1` for `buf2[0]`.

I'm happy to just delete the stats computation if we land on a suite of perftests that always have zero variance. For now, I'm leaning on that to have a sense of how deterministic the tests are :)

Reply:
Resolving — I have since deleted the stats computation. Let's go ahead with these hardcoded expectations on cycle counts until we see whether it's too flaky.